pkg/api/message/v1/service.go

```diff
@@ -195,61 +176,92 @@ func (s *Service) Subscribe(req *proto.SubscribeRequest, stream proto.MessageApi
 	log := s.log.Named("subscribe").With(zap.Strings("content_topics", req.ContentTopics))
 	log.Debug("started")
 	defer log.Debug("stopped")
+
 	// Send a header (any header) to fix an issue with Tonic based GRPC clients.
 	// See: https://github.com/xmtp/libxmtp/pull/58
 	_ = stream.SendHeader(metadata.Pairs("subscribed", "true"))
-	subC := s.dispatcher.Register(nil, req.ContentTopics...)
-	defer s.dispatcher.Unregister(subC)
-
-	for {
-		select {
-		case <-stream.Context().Done():
-			log.Debug("stream closed")
-			return nil
-		case <-s.ctx.Done():
-			log.Info("service closed")
-			return nil
-		case obj := <-subC:
-			env, ok := obj.(*proto.Envelope)
-			if !ok {
-				log.Warn("non-envelope received on subscription channel", zap.Any("object", obj))
-				continue
-			}
 
-			err := stream.Send(env)
+	var streamLock sync.Mutex
+	for _, topic := range req.ContentTopics {
+		sub, err := s.nc.Subscribe(topic, func(msg *nats.Msg) {
+			var env proto.Envelope
+			err := pb.Unmarshal(msg.Data, &env)
 			if err != nil {
-				log.Error("sending envelope to subscriber", zap.Error(err))
+				log.Error("parsing envelope from bytes", zap.Error(err))
+				return
 			}
+			if topic == natsWildcardTopic && !isValidSubscribeAllTopic(env.ContentTopic) {
+				return
+			}
+			func() {
```

Comment:
Is this just being wrapped in an anonymous function so we can use defer in it?

Reply:
Yep exactly